### PR TITLE
Sentinel middleware support

### PR DESCRIPTION
**Pull request by beiwei30**
# Description

_Please explain the changes you've made_

Integrate [sentinel](https://github.com/alibaba/sentinel-golang) as Dapr's middleware

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #828 

#### sample component configuration

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: sentinel
spec:
  type: middleware.http.sentinel
  version: v1
  metadata:
    - name: appName
      value: "nodeapp"
    - name: logDir
      value: "/var/tmp"
    - name: flowRules
      value: >-
        [
        	{
        		"resource": "POST:/v1.0/invoke/nodeapp/method/neworder",
        		"threshold": 10,
        		"tokenCalculateStrategy": 0,
        		"controlBehavior": 0
        	}
        ]
```

#### sample dapr config

```yaml
apiVersion: dapr.io/v1alpha1
kind: Configuration
metadata:
  name: daprConfig
spec:
  httpPipeline:
    handlers:
    - name: sentinel
      type: middleware.http.sentinel
```

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: middleware/http/sentinel/logger.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package sentinel
+
+import (
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/dapr/dapr/pkg/logger"
```
Comment (pkedy) [inline middleware/http/sentinel/logger.go:5]: This has moved to `"github.com/dapr/kit/logger"`

File: middleware/http/sentinel/middleware.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+
+	sentinel "github.com/alibaba/sentinel-golang/api"
+	"github.com/alibaba/sentinel-golang/core/base"
+	"github.com/alibaba/sentinel-golang/core/config"
+	"github.com/dapr/components-contrib/middleware"
+	"github.com/dapr/dapr/pkg/logger"
```
Comment (pkedy) [inline middleware/http/sentinel/middleware.go:16]: This has moved to `"github.com/dapr/kit/logger"`

File: middleware/http/sentinel/middleware_test.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
+package sentinel
+
+import (
+	"testing"
+
+	"github.com/dapr/components-contrib/middleware"
+	"github.com/dapr/dapr/pkg/logger"
```
Comment (pkedy) [inline middleware/http/sentinel/middleware_test.go:7]: This has moved to `"github.com/dapr/kit/logger"`

**Review by pkedy — CHANGES_REQUESTED** [review]
"github.com/dapr/dapr/pkg/logger" moved to "github.com/dapr/kit/logger" in order to eliminate a circular dependency.

Please change these imports and re-run `go mod tidy` using Go 1.16.

File: middleware/http/sentinel/middleware.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		return func(ctx *fasthttp.RequestCtx) {
+			resourceName := string(ctx.Method()) + ":" + string(ctx.Path())
+			entry, err := sentinel.Entry(
+				resourceName,
+				sentinel.WithResourceType(base.ResTypeWeb),
+				sentinel.WithTrafficType(base.Inbound),
+			)
+
+			if err != nil {
+				ctx.Error(fasthttp.StatusMessage(fasthttp.StatusTooManyRequests), fasthttp.StatusTooManyRequests)
```
Comment (pkedy) [inline middleware/http/sentinel/middleware.go:76]: I am not familiar with Sentinel, but this status code here seems strange. Should it be InternalServerError instead?

File: go.mod [hunk 5]
```diff
@@ -18,33 +18,32 @@ require (
...
 	github.com/apache/pulsar-client-go v0.1.0
 	github.com/apache/thrift v0.14.0 // indirect
 	github.com/aws/aws-sdk-go v1.27.0
-	github.com/baiyubin/aliyun-sts-go-sdk v0.0.0-20180326062324-cfa1a18b161f // indirect
 	github.com/bradfitz/gomemcache v0.0.0-20190913173617-a41fca850d0b
 	github.com/cenkalti/backoff/v4 v4.1.0
 	github.com/coreos/go-oidc v2.1.0+incompatible
 	github.com/cyphar/filepath-securejoin v0.2.2
 	github.com/dancannon/gorethink v4.0.0+incompatible
+	github.com/dapr/dapr v1.1.1
```
Comment (pkedy) [inline go.mod:36]: components-contrib should no longer depend on dapr/dapr

**Review by pkedy — COMMENTED** [review]
Just a question about a status code being returned

File: middleware/http/sentinel/middleware.go (hunk 4 above)
Comment (beiwei30) [reply middleware/http/sentinel/middleware.go:76]: I think it's fine. Pls. refer https://github.com/alibaba/sentinel-golang/blob/master/pkg/adapters/gin/middleware.go#L34. It uses the same error code. 

I will invite the authors of Sentinel to drop a comment against this too.

File: middleware/http/sentinel/logger.go (hunk 1 above)
Comment (beiwei30) [reply middleware/http/sentinel/logger.go]: fixed.

File: middleware/http/sentinel/middleware.go (hunk 2 above)
Comment (beiwei30) [reply middleware/http/sentinel/middleware.go]: fixed.

File: middleware/http/sentinel/middleware_test.go (hunk 3 above)
Comment (beiwei30) [reply middleware/http/sentinel/middleware_test.go]: fixed.

File: go.mod (hunk 5 above)
Comment (beiwei30) [reply go.mod]: fixed.

File: middleware/http/sentinel/middleware.go (hunk 4 above)
Comment (sczyh30) [reply middleware/http/sentinel/middleware.go:76]: The 429 status code has semantics for rate-limiting. Refer https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429

> The HTTP 429 Too Many Requests response status code indicates the user has sent too many requests in a given amount of time ("rate limiting").

Comment (beiwei30) [reply middleware/http/sentinel/middleware.go:76]: Correct. I confirmed with @sczyh30, the Sentinel's maintainer. With regarding to HTTP scenario, it is intentional to use "Too Many Requests" for all possible errors returned from Sentinel.

**Comment by beiwei30** [comment]
@pkedy would you mind to take a look again?

File: middleware/http/sentinel/middleware.go (hunk 4 above)
Comment (pkedy) [reply middleware/http/sentinel/middleware.go:76]: Thanks @beiwei30 and @sczyh30 for confirming!

**Review by pkedy — DISMISSED** [review]

File: middleware/http/sentinel/rules.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	return newDataSource(rules, datasource.NewHotSpotParamRulesHandler(datasource.HotSpotParamRuleJsonArrayParser))
+}
+
+func newIsolationRuleDataSource(rules string) (datasource.DataSource, error) {
+	return newDataSource(rules, datasource.NewIsolationRulesHandler(datasource.IsolationRuleJsonArrayParser))
+}
+
+func newSystemRuleDataSource(rules string) (datasource.DataSource, error) {
+	return newDataSource(rules, datasource.NewSystemRulesHandler(datasource.SystemRuleJsonArrayParser))
+}
```
Comment (halspang) [inline middleware/http/sentinel/rules.go:52]: Any reason the system rules aren't included in the rules_test?

File: middleware/http/sentinel/middleware.go [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+				return
+			}
+
+			defer entry.Exit()
+			h(ctx)
+		}
+	}, nil
+}
+
+func (m *Middleware) loadSentinelRules(meta *middlewareMetadata) error {
```
Comment (halspang) [inline middleware/http/sentinel/middleware.go:87]: Is there a good way to test the rules? Both in terms of if they load correctly/incorrectly and if the rules are being applied correctly?

**Review by halspang — COMMENTED** [review]
Just some questions around testing.

**Review by artursouza — COMMENTED** [review]
Please, see comments from @halspang

File: middleware/http/sentinel/rules.go (hunk 6 above)
Comment (beiwei30) [reply middleware/http/sentinel/rules.go:52]: add test against system rule.

File: middleware/http/sentinel/middleware.go (hunk 7 above)
Comment (beiwei30) [reply middleware/http/sentinel/middleware.go:87]: added a new test for `loadSentinelRules`

**Comment by beiwei30** [comment]
@halspang I've already revised this PR, pls. check again.

**Review by halspang — APPROVED** [review]

**Comment by beiwei30** [comment]
@pkedy @halspang I have resolved conflicts, pls. check again.

**Comment by beiwei30** [comment]
@pkedy @halspang, any update on this?

**Review by halspang — APPROVED** [review]

**Review by pkedy — APPROVED** [review]

**Review by artursouza — APPROVED** [review]

**Pull request merged**